### PR TITLE
[ownership] Make checked_cast_br, destructure_struct, and destructure_tuple real forwarding instructions

### PR DESCRIPTION
**Pull request by gottesmm**
More changes so that we can eliminate the OperandOwnershipKindMap.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
Turns out this is dependent on #34633

File: lib/SIL/Verifier/SILVerifier.cpp [hunk 1]
```diff
@@ -2915,6 +2915,41 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
     require(!sd->isResilient(F.getModule().getSwiftModule(),
                              F.getResilienceExpansion()),
             "cannot access storage of resilient struct");
+    if (F.hasOwnership()) {
+      // Make sure that all of our destructure results ownership kinds are
+      // compatible with our destructure_struct's ownership kind /and/ that if
+      // our destructure ownership kind is non-trivial then all non-trivial
+      // results must have the same ownership kind as our operand.
+      auto parentKind = DSI->getOwnershipKind();
+      for (const DestructureStructResult &result : DSI->getAllResultsBuffer()) {
```
Comment (atrick) [inline lib/SIL/Verifier/SILVerifier.cpp:2924]: What you're really doing here is using the parent ownership to constrain the result ownership. Now that you can convert between ValueOwnershipKind and OwnershipConstraint you can express these requirements as a single requirement:
`require(OwnershipConstraint(parentKind).satisfiedBy(result.getOwnershipKind())`
or, alternatively you can just say 
`require(parentKind.merge(result.getOwnershipKind()) == parentKind)`

**Review by atrick — APPROVED** [review]
Looks great, but I don't think `isCompatibleWith` is the right API to use for checking ownership consistency and you shouldn't need to special-case 'None'

File: lib/SIL/Verifier/SILVerifier.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SIL/Verifier/SILVerifier.cpp:2924]: Right, I need to finish refactoring though.

I think that parentKind.isCompatibleWith is actually correct and my extra requires are unnecessary. I was worried at the time about a .none parent ownership being compatible with a result that is not .none. But that can't happen. I am going to remove that part.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
Linux failure wasn't this commit. Was a missing module.

**Comment by gottesmm** [comment]
@swift-ci test linux platform

**Comment by gottesmm** [comment]
macOS is going to fail the same way. I am going to wait though, let it finish running /all/ the Swift tests up to that build breakage in tsc. And then I am going to smoke test.

**Comment by gottesmm** [comment]
@swift-ci smoke test OS X platform

**Comment by gottesmm** [comment]
OS X failed as expected. All the tests before swiftpm passed.

**Pull request merged**